include/dali/core/mm/async_pool.h

```diff
@@ -0,0 +1,431 @@
+    size_t          alignment = alignof(std::max_align_t);
+    CUDAEvent       event;
+    bool            is_ready = false;
+    pending_free   *prev = nullptr, *next = nullptr;
+
+    bool ready() {
+      if (is_ready)
+        return true;
+      cudaError_t e = cudaEventQuery(event);
+      switch (e) {
+        case cudaSuccess:
+          return is_ready = true;
+        case cudaErrorNotReady:
+          return false;
+        case cudaErrorCudartUnloading:
+          cudaGetLastError();
+          return true;
+        default:
+          cudaGetLastError();
+          throw CUDAError(e);
+      }
+    }
+  };
+
+  struct PendingFreeList {
+    pending_free *head = nullptr, *tail = nullptr;
+  };
+
+  /// @brief Non-coalescing free blocks deallocated on a stream
+  struct PerStreamFreeBlocks {
+    using size_pending = std::pair<size_t, pending_free *>;
+
+    PerStreamFreeBlocks() = default;
+    PerStreamFreeBlocks(const PerStreamFreeBlocks &) = delete;
+
+    /// @brief A map that associates block size with the free block; used for best-fit allocation
+    detail::pooled_set<size_pending, true> by_size;
+
+    /**
+     * @brief List of free blocks, stored in deallocation order, head being most recent.
+     *
+     * The items in the list are kept in the order there were deallocated. Since we're using
+     * stream-ordered deallocation and all list items were deallocated on the same stream, this
+     * list has a well defined transition point from pending to complete deallocations - this
+     * is exploited when returning complete deallocations to the global pool.
+     */
+    PendingFreeList free_list;
+  };
+
+  /**
+   * @brief Try to allocate a block from per-stream free memory
+   *
+   * If the allocation fails, the function returns nullptr.
+   */
+  void *try_allocate(PerStreamFreeBlocks &from, size_t bytes, size_t alignment) {
+    unsigned max_padding = std::min<size_t>(std::max<size_t>(bytes / 16, 16), (1 << 20));
+    const unsigned remainder_alignment = 16;
+    const unsigned min_split_remainder = 16;
+    for (auto it = from.by_size.lower_bound({ bytes, nullptr }); it != from.by_size.end(); ++it) {
+      size_t block_size = it->first;
+      pending_free *f = it->second;
+      char *base = f->addr;
+      char *aligned = detail::align_ptr(base, alignment);
+      size_t front_padding = aligned - base;
+      assert(static_cast<ptrdiff_t>(front_padding) >= 0);
+      // NOTE: block_size - front_padding >= size  can overflow and fail - meh, unsigned size_t
+      if (block_size >= bytes + front_padding) {
+        if (!supports_splitting && block_size - bytes - front_padding > max_padding)
+          return nullptr;  // no point in continuing - the map is sorted
+
+        from.by_size.erase(it);
+        char *block_end = base + block_size;
+        char *remainder = detail::align_ptr(aligned + bytes, remainder_alignment);
+        bool split = supports_splitting && remainder + min_split_remainder < block_end;
```

Comment:
```suggestion
        bool split = supports_splitting && min_split_remainder  < block_end - remainder ;
```
It is the same but is is cleared that the remaining part should be greater than `min_split_remainder`.

Reply:
It is not the same because size_t is unsigned and it would wrap around into huge regions. Been there, debugged that.